### PR TITLE
Add INT16 support in modify_model_interface

### PR DESCRIPTION
**Pull request by TamasArm**
* Added support for 16bit quantized models in modify_model_interface.cc
 * Added extra arguments for inputing 16bit quantized models in modify_model_interface_main.cc
 * Modified 2 of the unit tests to be parametrized and covered 16bit models in modify_model_interface_test.cc

### CODE REVIEW

**Comment by akarmi** [comment]
Related to #36632

**Comment by gbaned** [comment]
@TamasArm Thank you for your contribution. Can you please sign CLA? Thanks!

**Comment by MatteoArm** [comment]
@gbaned

**Comment by TamasArm** [comment]
@googlebot I signed it!

**Comment by gbaned** [comment]
@TamasArm Can you please resolve conflicts? Thanks!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 1]
```diff
@@ -46,7 +46,8 @@ struct TensorOpTensor {
 // Finds float tensors that are model inputs and is consumed by a quantize Op.
 // The returned TensorOpTensor should have reverse order.
 std::vector<TensorOpTensor> GetInputTensors(ModelT* model,
-                                            ErrorReporter* error_reporter) {
+                                            ErrorReporter* error_reporter,
+                                            const TensorType& input_type) {
```
Comment (renjie-liu) [inline tensorflow/lite/tools/optimize/modify_model_interface.cc:50]: nit: prefer move immutable arguments before mutable arguments

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 2]
```diff
@@ -85,10 +86,16 @@ std::vector<TensorOpTensor> GetInputTensors(ModelT* model,
...
-      if (quant_output->type != TensorType_INT8) {
+      if (quant_output->type != TensorType_INT8 && quant_output->type != TensorType_INT16) {
         TF_LITE_REPORT_ERROR(error_reporter,
-                             "modify_model_interface currently only support "
-                             "int8 quantized models.");
+                             "modify_model_interface currently only supports "
+                             "int8 and int16 quantized models.");
+      }
+      if (quant_output->type != input_type) {
+        if (!(quant_output->type == TensorType_INT8 && input_type == TensorType_UINT8)) {
```
Comment (renjie-liu) [inline tensorflow/lite/tools/optimize/modify_model_interface.cc:95]: this deserves a comment

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 3]
```diff
@@ -103,7 +110,8 @@ std::vector<TensorOpTensor> GetInputTensors(ModelT* model,
 // Finds float tensors that are model output and is consumed by a dequantize Op.
 // The returned TensorOpTensor should have reverse order.
 std::vector<TensorOpTensor> GetOutputTensors(ModelT* model,
-                                             ErrorReporter* error_reporter) {
+                                             ErrorReporter* error_reporter,
+                                             const TensorType& output_type) {
```
Comment (renjie-liu) [inline tensorflow/lite/tools/optimize/modify_model_interface.cc:114]: ditto

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 4]
```diff
@@ -288,9 +302,9 @@ std::unique_ptr<tflite::ModelT> CreateMutableModelFromFile(
   return copied_model;
 }
 
-int GetOriginalNumberOfTensors(ModelT* model, ErrorReporter* error_reporter) {
-  std::vector<TensorOpTensor> outputs = GetOutputTensors(model, error_reporter);
-  std::vector<TensorOpTensor> inputs = GetInputTensors(model, error_reporter);
+int GetOriginalNumberOfTensors(ModelT* model, ErrorReporter* error_reporter, const TensorType& input_type, const TensorType& output_type) {
```
Comment (renjie-liu) [inline tensorflow/lite/tools/optimize/modify_model_interface.cc:305]: can you clang format this one? thanks

File: tensorflow/lite/tools/optimize/modify_model_interface.cc (hunk 2 above)
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc:98]: done, thank you!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc (hunk 1 above)
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: done, thank you!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc (hunk 3 above)
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: done, thank you!

**Review by MeghnaNatraj — COMMENTED** [review]
Thanks for adding support for INT16 quantized models!

Apart from the requested changes, if possible
1. Improve error description: Describe the issue with actual and expected types.
2. In the comments and error descriptions which you have updated, consistently use uppercase or lowercase for type names eg: Use either 'uint8' or 'UINT8 '

To keep this CL minimal, apply these changes *only* to code that you have/need to update.

File: tensorflow/lite/tools/optimize/modify_model_interface.h [hunk 5]
```diff
@@ -24,7 +24,7 @@ namespace optimize {
 // Changes the interface of a quantized model. This method allows the users to
 // replace float interface with other types.
 // This populates the builder with the new model.
-// Currently only int8 and unit8 are supported.
+// Currently only int8, int16 and unit8 are supported.
```
[the remark this answers is not in the document]
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.h]: Done, thank you!

File: tensorflow/lite/tools/optimize/modify_model_interface_main.cc [hunk 6]
```diff
@@ -25,24 +25,21 @@ int main(int argc, char** argv) {
...
-    input = tflite::TensorType_UINT8;
-  }
-  if (!strcmp(argv[4], "uint8")) {
-    output = tflite::TensorType_UINT8;
+  try {
+    input = supported_types.at(argv[3]);
+    output = supported_types.at(argv[4]);
+  } catch (const std::out_of_range&) {
+    printf(
+        "Only supports uint8, int8 and int16 for input and output interfaces");
```
[the remark this answers is not in the document]
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface_main.cc]: Done, thank you!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 7]
```diff
@@ -85,10 +86,23 @@ std::vector<TensorOpTensor> GetInputTensors(ModelT* model,
...
+      // ones we're trying to remove.
+      if (quant_output->type != input_type) {
+        // An exception from this is when we are setting the input or output
+        // type to UINT8.
+        if (!(quant_output->type == TensorType_INT8 &&
+              input_type == TensorType_UINT8)) {
+          TF_LITE_REPORT_ERROR(
+              error_reporter,
+              "Model's type incompatible with output type argument.");
+        }
```
[the remark this answers is not in the document]
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: Done, thanks!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 8]
```diff
@@ -357,13 +396,26 @@ TfLiteStatus ModifyModelInterface(const string& input_file,
...
-    output_override_type = tflite::TensorType_UINT8;
+  switch (output_type) {
+    case tflite::TensorType_UINT8:
+    case tflite::TensorType_INT8:
+    case tflite::TensorType_INT16:
+      output_override_type = output_type;
+      break;
+    default:
+      return kTfLiteError;
   }
```
[the remark this answers is not in the document]
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: Done, thanks!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 9]
```diff
@@ -142,13 +157,22 @@ std::vector<TensorOpTensor> GetOutputTensors(ModelT* model,
...
         return {};
       }
+      if (dequant_input->type != output_type) {
+        if (!(dequant_input->type == TensorType_INT8 &&
+              output_type == TensorType_UINT8)) {
+          TF_LITE_REPORT_ERROR(
+              error_reporter,
+              "Model's type incompatible with output type argument.");
+        }
+      }
```
[the remark this answers is not in the document]
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: Done, thanks!

**Comment by TamasArm** [comment]
Is there anything else I can do to help move this forward?

**Comment by yyoon** [comment]
Gentle ping: @MeghnaNatraj @renjie-liu

**Review by renjie-liu — COMMENTED** [review]
sorry for the delay!

looks good, but please wait for others' comment.

thanks!

File: tensorflow/lite/tools/optimize/modify_model_interface.cc [hunk 10]
```diff
@@ -18,9 +18,10 @@ limitations under the License.
 #include <memory>
 #include <sstream>
 #include <unordered_set>
+#include <typeinfo>
```
Comment (jdduke) [inline tensorflow/lite/tools/optimize/modify_model_interface.cc:21]: Why is this necessary?

**Comment by MeghnaNatraj** [comment]
@TamasArm could you merge the latest changes? There is just one merge conflict. 
Otherwise this looks good to me as well!

@jianlijianli Can you take a final look at this before we merge these changes?

File: tensorflow/lite/tools/optimize/modify_model_interface.cc (hunk 10 above)
Comment (MeghnaNatraj) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc:21]: +1. 

I don't see any usage of this function eg: typeid() function

Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface.cc]: You are right, removed it, thank you!

**Comment by TamasArm** [comment]
Bear with me, I am trying to fix the issues with extra commits on my log after a failed rebase, sorry about that!

**Comment by TamasArm** [comment]
Fixed now, it's free to review again!

**Review by akarmi — CHANGES_REQUESTED** [review]
Should we also be making changes in the corresponding [Python binding](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.py)?

**Review by MeghnaNatraj — APPROVED** [review]

**Comment by MeghnaNatraj** [comment]
@akarmi Yes, we should be making the changes in the python equivalent code as well. 

@TamasArm If you're not familiar with this, let us know and I can make this change later.

If you're interested to try: We have a python wrapper that internally invokes the c++ library `modify_model_interface.cc` which you have updated. The workflow is: User runs [modify_model_interface.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.py) --> [modify_model_interface_lib.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface_lib.py) --> [pybind (python to c++ connection) file](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface.cc) which ultimately invokes `modify_model_interface.cc` 

To support your changes in python, you need to update [modify_model_interface_constants.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface_constants.py) to add INT16 and a test to create an INT16 quantized model and validate this in [modify_model_interface_lib_test.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tools/optimize/python/modify_model_interface_lib_test.py)

If this is confusing, do let us know and we can update it on our end :)

**Comment by TamasArm** [comment]
@MeghnaNatraj Thank you for the explanation! It seems clear to me, I'll get started on it right away, I'll try to be done with it by Monday night. Should I do it in a seperate PR?

**Comment by MeghnaNatraj** [comment]
@TamasArm Yes, let us add this change in a separate PR. I'll go ahead and work on merging this change.

File: tensorflow/lite/tools/optimize/modify_model_interface_test.cc [hunk 11]
```diff
@@ -396,8 +425,45 @@ TEST(ModelInterface, MixedTypeSingleInputOutput) {
...
+  auto output = model->subgraphs[0]->tensors[output_quant_op->outputs[0]].get();
+  EXPECT_EQ(output->name, "tensor_3");
+  EXPECT_EQ(output->type, TensorType_UINT8);
+  EXPECT_FLOAT_EQ(output->quantization->scale[0], 0.12);
+  EXPECT_EQ(output->quantization->zero_point[0], 178);
+}
+
 TEST(ModelInterface, Uint8MutipleInputOutput) {
-  auto model = CreateQuantizedModelMultipleInputOutput();
+  auto model = CreateQuantizedModelMultipleInputOutput(TensorType_INT8);
```
Comment (rthadur) [inline tensorflow/lite/tools/optimize/modify_model_interface_test.cc:466]: `tensorflow/lite/tools/optimize/modify_model_interface_test.cc:522:16: error: no matching function for call to 'CreateQuantizedModelMultipleInputOutput'
  auto model = CreateQuantizedModelMultipleInputOutput();`

**Review by rthadur — CHANGES_REQUESTED** [review]

**Review by MeghnaNatraj — DISMISSED** [review]

**Review by rthadur — APPROVED** [review]

File: tensorflow/lite/tools/optimize/modify_model_interface_test.cc (hunk 11 above)
Comment (TamasArm) [reply tensorflow/lite/tools/optimize/modify_model_interface_test.cc:466]: Thank you, addressed it in the latest commit.
